### PR TITLE
[KYUUBI #6253] Support running JDBC engine on YARN AM

### LINKED ISSUE #6253: [TASK][MEDIUM] Support running JDBC engine on YARN AM
### What's the level of this task?

MEDIUM

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before creating

- [X] I have searched in the [task list](https://github.com/orgs/apache/projects/343) and found no similar tasks.

### Mentor

- [X] I have sufficient expertise on this task, and I volunteer to be a mentor of this task to guide contributors through the task.

### Skill requirements

- Knowledge of Scala, Kyuubi architecture, Apache YARN
- Kerberos and Hadoop Security (optional)

### Background and Goals

Kyuubi supports running the Hive engine on YARN AM (Application Master) since 1.9.0, see https://github.com/apache/kyuubi/pull/5868 and https://github.com/apache/kyuubi/pull/6199, it removes the requirement of running engine process on Kyuubi Server's machine but moves to YARN cluster, just like Spark on YARN cluster mode and Flink YARN application mode.

We'd like to see similar functionalities for the JDBC engine.

### Implementation steps

Refer to the Hive engine implementation to complete the code, test, CI, etc.

### Additional context

Introduction of [2024H1 Kyuubi Code Contribution Program](https://github.com/apache/kyuubi/issues/6232)

This may help for testing https://github.com/awesome-kyuubi/hadoop-testing

### PR DESCRIPTION
**Pull request by lsm1**
# :mag: Description
## Issue References 🔗

This pull request fixes #6253 

## Describe Your Solution 🔧

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Behavior With This Pull Request :tada:

image

#### Related Unit Tests

---

# Checklist 📝

- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala [hunk 1]
```diff
@@ -315,8 +315,10 @@ abstract class EngineYarnModeSubmitter extends Logging {
       }
       // respect the following priority loading configuration, and distinct files
       // hive configuration -> hadoop configuration -> yarn configuration
-      val hiveConf = kyuubiConf.getOption(KYUUBI_ENGINE_DEPLOY_YARN_MODE_HIVE_CONF_KEY)
-      listDistinctFiles(hiveConf.get).foreach(putEntry)
+      if ("hive".equalsIgnoreCase(engineType)) {
```
Comment (Yikf) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala:318]: For this class, we want it to be as generic as possible. Can it be better abstracted? This would allow different subclasses to implement varying logic.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitter.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  override var engineType: String = "jdbc"
+
+  override def engineMainClass(): String = JdbcSQLEngine.getClass.getName
+
+  /**
+   * Jar list for the Hive engine.
+   */
+  override def engineExtraJars(): Seq[File] = {
+    val hadoopCp = sys.env.get("HIVE_HADOOP_CLASSPATH")
```
Comment (Yikf) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitter.scala:55]: nit

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitter.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    val extraCp = kyuubiConf.get(ENGINE_JDBC_EXTRA_CLASSPATH)
+    val jars = new ListBuffer[File]
+    hadoopCp.foreach(cp => parseClasspath(cp, jars))
+    extraCp.foreach(cp => parseClasspath(cp, jars))
+    jars.toSeq
+  }
+
+  private[jdbc] def parseClasspath(classpath: String, jars: ListBuffer[File]): Unit = {
```
Comment (Yikf) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitter.scala:63]: Should this be a public function?

File: externals/kyuubi-jdbc-engine/src/test/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitterSuite.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+package org.apache.kyuubi.engine.jdbc.deploy
+
+import java.io.File
+
+import scala.collection.mutable.ListBuffer
+
+import org.apache.kyuubi.{KYUUBI_VERSION, KyuubiFunSuite, SCALA_COMPILE_VERSION}
+import org.apache.kyuubi.util.JavaUtils
+
+class JdbcYarnModeSubmitterSuite extends KyuubiFunSuite {
```
Comment (Yikf) [inline externals/kyuubi-jdbc-engine/src/test/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitterSuite.scala:26]: If those functions become public, this suite will no longer be necessary.

File: integration-tests/kyuubi-jdbc-it/src/test/scala/org/apache/kyuubi/it/jdbc/doris/WithKyuubiServerAndDorisContainerYarnMode.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+import java.nio.file.{Files, Path, Paths}
+
+import org.apache.kyuubi.WithKyuubiServerAndHadoopMiniCluster
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf.{ENGINE_IDLE_TIMEOUT, ENGINE_JDBC_EXTRA_CLASSPATH, ENGINE_TYPE, KYUUBI_ENGINE_ENV_PREFIX, KYUUBI_HOME}
+import org.apache.kyuubi.engine.deploy.DeployMode
+import org.apache.kyuubi.engine.jdbc.doris.WithDorisEngine
+
+trait WithKyuubiServerAndDorisContainerYarnMode extends WithKyuubiServerAndHadoopMiniCluster
```
Comment (Yikf) [inline integration-tests/kyuubi-jdbc-it/src/test/scala/org/apache/kyuubi/it/jdbc/doris/WithKyuubiServerAndDorisContainerYarnMode.scala:28]: Perhaps we only need to validate one JDBC engine? Because the logic for yarn mode is the same.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcProcessBuilder.scala [hunk 6]
```diff
@@ -113,3 +116,55 @@ class JdbcProcessBuilder(
...
+      conf: KyuubiConf,
+      engineRefId: String,
+      extraEngineLog: Option[OperationLog],
+      defaultEngineName: String): JdbcProcessBuilder = {
+    checkKeytab(proxyUser, conf)
+    DeployMode.withName(conf.get(ENGINE_JDBC_DEPLOY_MODE)) match {
+      case LOCAL =>
+        new JdbcProcessBuilder(proxyUser, doAsEnabled, conf, engineRefId, extraEngineLog)
+      case YARN =>
+        warn(s"Hive on YARN model is experimental.")
```
Comment (Yikf) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcProcessBuilder.scala:136]: nit: Hive -> JDBC Engine

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcYarnModeProcessBuilder.scala [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf.{ENGINE_JDBC_EXTRA_CLASSPATH, ENGINE_JDBC_MEMORY}
+import org.apache.kyuubi.config.KyuubiReservedKeys.{KYUUBI_ENGINE_ID, KYUUBI_SESSION_USER_KEY}
+import org.apache.kyuubi.engine.{ApplicationManagerInfo, KyuubiApplicationManager}
+import org.apache.kyuubi.engine.deploy.yarn.EngineYarnModeSubmitter._
+import org.apache.kyuubi.operation.log.OperationLog
+import org.apache.kyuubi.util.command.CommandLineUtils.{confKeyValue, confKeyValues}
+
+/**
+ * A process builder for Hive on Yarn.
```
Comment (Yikf) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcYarnModeProcessBuilder.scala:36]: nit. You can check overall if the ported code has been modified from Hive to JDBC.

File: integration-tests/kyuubi-jdbc-it/src/test/scala/org/apache/kyuubi/it/jdbc/doris/WithKyuubiServerAndDorisContainerYarnMode.scala (hunk 5 above)
Comment (lsm1) [reply integration-tests/kyuubi-jdbc-it/src/test/scala/org/apache/kyuubi/it/jdbc/doris/WithKyuubiServerAndDorisContainerYarnMode.scala:28]: ok, just keep mysql suite

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala [hunk 8]
```diff
@@ -315,13 +315,7 @@ abstract class EngineYarnModeSubmitter extends Logging {
       }
       // respect the following priority loading configuration, and distinct files
       // hive configuration -> hadoop configuration -> yarn configuration
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala:317]: the comments should be updated and placed into right place(s)

**Review by pan3793 — APPROVED** [review]

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcSQLEngine.scala [hunk 9]
```diff
@@ -66,15 +72,45 @@ object JdbcSQLEngine extends Logging {
...
+          UserGroupInformation.getCurrentUser
+        case DeployMode.LOCAL if proxyUser.get != realUser.getShortUserName =>
+          val newUGI = UserGroupInformation.createProxyUser(proxyUser.get, realUser)
+          newUGI.doAs(new PrivilegedExceptionAction[Unit] {
+            override def run(): Unit = {
+              val engineCredentials =
+                kyuubiConf.getOption(KyuubiReservedKeys.KYUUBI_ENGINE_CREDENTIALS_KEY)
+              kyuubiConf.unset(KyuubiReservedKeys.KYUUBI_ENGINE_CREDENTIALS_KEY)
+              engineCredentials.filter(_.nonEmpty).foreach { credentials =>
+              }
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcSQLEngine.scala:98]: why keep a empty body here?

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitter.scala [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    Utils.fromCommandLineArgs(args, kyuubiConf)
+
+    if (UserGroupInformation.isSecurityEnabled) {
+      require(
+        kyuubiConf.get(KyuubiConf.ENGINE_PRINCIPAL).isDefined
+          && kyuubiConf.get(KyuubiConf.ENGINE_KEYTAB).isDefined,
+        s"${KyuubiConf.ENGINE_PRINCIPAL.key} and " +
+          s"${KyuubiConf.ENGINE_KEYTAB.key} must be set when submit " +
+          s"${JdbcSQLEngine.getClass.getSimpleName.stripSuffix("$")} to YARN")
+    }
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/deploy/JdbcYarnModeSubmitter.scala:43]: actually, this condition could be relaxed. but this can be done later, independently. cc @cxzl25 

for hive, we require keytab to run MR jobs, but for JDBC, most DB does not require kerberos authentication.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcSQLEngine.scala [hunk 11]
```diff
@@ -66,15 +72,45 @@ object JdbcSQLEngine extends Logging {
 
   def main(args: Array[String]): Unit = {
     SignalRegister.registerLogger(logger)
-
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcSQLEngine.scala:69]: avoid unnecessary change

File: externals/kyuubi-hive-sql-engine/src/main/scala/org/apache/kyuubi/engine/hive/deploy/HiveYarnModeSubmitter.scala [hunk 12]
```diff
@@ -60,18 +61,10 @@ object HiveYarnModeSubmitter extends EngineYarnModeSubmitter {
...
-            f.listFiles().filter(_.getName.endsWith(".jar")).foreach(jars += _)
-          case _ =>
-        }
-      } else {
-        jars += new File(cp)
-      }
-    }
+  override def listConfFiles(): Seq[File] = {
+    // respect the following priority loading configuration, and distinct files
+    // hive configuration -> hadoop configuration -> yarn configuration
```
Comment (pan3793) [inline externals/kyuubi-hive-sql-engine/src/main/scala/org/apache/kyuubi/engine/hive/deploy/HiveYarnModeSubmitter.scala:66]: please supply similar message on the super method

**Comment by pan3793** [comment]
Just minor comments, LGTM

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcSQLEngine.scala [hunk 13]
```diff
@@ -88,17 +89,7 @@ object JdbcSQLEngine extends Logging {
...
-            override def run(): Unit = {
-              val engineCredentials =
-                kyuubiConf.getOption(KyuubiReservedKeys.KYUUBI_ENGINE_CREDENTIALS_KEY)
-              kyuubiConf.unset(KyuubiReservedKeys.KYUUBI_ENGINE_CREDENTIALS_KEY)
-              engineCredentials.filter(_.nonEmpty).foreach { credentials =>
-              }
-            }
-          })
-          newUGI
+          UserGroupInformation.createProxyUser(proxyUser.get, realUser)
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/JdbcSQLEngine.scala:92]: `kyuubiConf.unset(KyuubiReservedKeys.KYUUBI_ENGINE_CREDENTIALS_KEY` is missed

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala [hunk 14]
```diff
@@ -352,6 +344,14 @@ abstract class EngineYarnModeSubmitter extends Logging {
     }.toSeq
   }
 
+  def listConfFiles(): Seq[File] = {
+    // respect the following priority loading configuration, and distinct files
+    // hive configuration -> hadoop configuration -> yarn configuration
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala:349]:
```suggestion
    // hadoop configuration -> yarn configuration
```

**Review by yikf — APPROVED** [review]
Overall, LGTM, but I would also like @zhouyifan279  to take a look at the Kerberos authentication-related code if you find a time.

**Review by cxzl25 — APPROVED** [review]

**Comment by cxzl25** [comment]
Thanks, merged to master.
